EIPS/eip-3520.md

```diff
@@ -0,0 +1,91 @@
+
+This EIP enables the development of new types of smart contracts as it can open new pathways for [EIP-721](./eip-721) NFTs and [EIP-20](./eip-20) tokens to detect which action their transaction is part of, such as detecting a liquidity provision to a decentralized exchange or a loan within a collateralized lending protocol.
+
+## Specification
+
+### ENTRYPOINT (`0x4a`)
+
+The `ENTRYPOINT` instruction uses 0 stack arguments and pushes the original `to` member of the transaction onto the stack. The address yielded by the instruction is a 160-bit value padded to 256-bits. The operation costs `G_base` to execute, similarly to `ORIGIN` (`0x32`).
+
+### `AUTHCALL` (`0xf7`) Interaction
+
+The [EIP-3074](./eip-3074) that will be part of the London fork has introduced a new call instruction called `AUTHCALL` (`0xf7`) that will replace a transaction's `ORIGIN` (`0x32`) with the context variable `authorized`. The intention of `AUTHCALL` is to prevent discrimination between smart contracts and EOAs which `ORIGIN` initially facilitated and as a result, it is sensible also replace the values retrieved by the `ENTRYPOINT` opcode to the target of an `AUTHCALL`.
+
+The `ENTRYPOINT` opcode by itself re-introduces discrimination into the system as it indirectly allows one to evaluate whether the smart contract code being executed is done so by an EOA by validating that `ENTRYPOINT == ADDRESS` where `ADDRESS` (`0x30`) retrieves the currently executing account address.
+
+In detail, the address returned by the `ENTRYPOINT` opcode will be equivalent to the `to` address specified in the `commit` of the `AUTH` message format. If a single-call `AUTH` message is defined, the opcode will retrieve the address of `to`. Otherwise, the opcode will retrieve the `to` address of each sub-call (i.e. `call 1`, `call 2` ... `call X`) depending on the call execution context.
+
+This interaction ensures full compatibility with [EIP-3074](./eip-3074) and ensures that no form of discrimination is introduced back into the system by this EIP.
+
+## Rationale
+
+### Naming Conventions
+
+The `ENTRYPOINT` instruction came to be by defining a sensible name that immediately and clearly depicts what it is meant to achieve by signaling the first interaction of a particular call, i.e. the entry-point.
+
+### Instruction Address Space
+
+Equivalent to [EIP-3508](./eip-3508).
+
+### Gas Cost
+
+The opcode ENTRYPOINT (`0x4a`) essentially performs the same thing as the opcode ORIGIN (`0x32`) and thus shares the exact same gas cost.
+
+### Dependency on EIP-3508
+
+The `ENTRYPOINT` (`0x4a`) instruction alone has no perceivable benefit as it can be replaced by the `AUTHCALL` (`0xf7`) instruction and as such should solely be introduced to the system in conjunction with its sister instructions, the `ORIGINDATA*` opcodes defined in [EIP-3508](./eip-3508). However, if a use case of `ENTRYPOINT` (`0x4a`) alone can be argued then this dependency will be omitted.
```

Comment:
Write EIPs in the final form, and change them if necessary based on new information during the review process.  Don't include things like "if X happens then this EIP will change" in the EIP itself.
```suggestion
The `ENTRYPOINT` (`0x4a`) instruction alone has no perceivable benefit as it can be replaced by the `AUTHCALL` (`0xf7`) instruction and as such should solely be introduced to the system in conjunction with the `ORIGINDATA*` opcodes defined in [EIP-3508](./eip-3508).
```